allennlp/data/iterators/bucket_iterator.py

```diff
@@ -0,0 +1,105 @@
+    and padding keys which the dataset will be sorted by before doing this batching,
+    causing inputs with similar length to be batched together, making computation
+    more efficient (as less time is wasted on padded elements of the batch).
+
+    Parameters
+    ----------
+    sorting_keys : List[Tuple[str, str]], optional (default = [])
+        To bucket inputs into batches, we want to group the instances by padding length, so that
+        we minimize the amount of padding necessary per batch. In order to do this, we need to
+        know which fields need what type of padding, and in what order.
+
+        For example:
+        [(sentence1, "sentence_length"), (sentence2, "sentence_length"), ("sentence1", "char_length")]
+
+        would sort a dataset first by the "sentence_length" of the "sentence1" field, then by the
+        "sentence_length" of the "sentence2" field, and finally by the "char_length" of the "sentence1"
+        field.
+        By default, the list of sorting keys is empty, meaning the dataset won't be sorted and batches
+        will just be padded using the max lengths of all fields requiring padding calculated per batch.
+    padding_noise: float, optional (default=.1)
+        When sorting by padding length, we add a bit of noise to the lengths, so that the sorting
+        isn't deterministic.  This parameter determines how much noise we add, as a percentage of
+        the actual padding value for each instance.
+    sort_every_epoch: bool, optional (default=True)
+        If ``True``, we will re-sort the data after every epoch, then re-group the instances into
+        batches.  If ``padding_noise`` is zero, this does nothing, but if it's non-zero, this will
+        give you a slightly different ordering, so you don't have exactly the same batches at every
+        epoch.  If you're doing adaptive batch sizes, this will lead to re-computing the adaptive
+        batches each epoch, which could give a different number of batches for the whole dataset,
+        which means each "epoch" might no longer correspond to `exactly` one pass over the data.
+        This is probably a pretty minor issue, though.
+    batch_size : int, optional, (default = 32)
+        The size of each batch of instances yielded when calling the iterator.
+    """
+
+    def __init__(self,
+                 sorting_keys: List[Tuple[str, str]] = None,
+                 padding_noise: float = 0.1,
+                 sort_every_epoch: bool = True,
+                 batch_size: int = 32):
+
+        self.sorting_keys = sorting_keys or []
+        self.padding_noise = padding_noise
+        self.sort_every_epoch = sort_every_epoch
+        super(BucketIterator, self).__init__(batch_size)
+
+    @overrides
+    def __call__(self, dataset: Dataset):
+        grouped_instances = self._create_batches(dataset)
+        self.last_num_batches = len(grouped_instances)  # pylint: disable=attribute-defined-outside-init
```

Comment:
Where is this attribute used?  It seems a little scary to assign something to `self` in `__call__` since a client has to remember the last time they called the iterator to know what the value of this attribute is?